gateleen-security/src/main/java/org/swisspush/gateleen/security/authorization/RoleAuthorizer.java

```diff
@@ -0,0 +1,217 @@
+        initialGrantedRoles.get(aclUriPattern).put("DELETE", new HashSet<>());
+        initialGrantedRoles.get(aclUriPattern).get("PUT").add(adminRole);
+        initialGrantedRoles.get(aclUriPattern).get("GET").add(adminRole);
+        initialGrantedRoles.get(aclUriPattern).get("DELETE").add(adminRole);
+
+        configUpdate();
+    }
+
+    public void handleIsAuthorized(final HttpServerRequest request, Future<Boolean> future) {
+        if (!isAuthorized(request)) {
+            ResponseStatusCodeLogUtil.info(request, StatusCode.FORBIDDEN, RoleAuthorizer.class);
+            request.response().setStatusCode(StatusCode.FORBIDDEN.getStatusCode());
+            request.response().setStatusMessage(StatusCode.FORBIDDEN.getStatusMessage());
+            request.response().end(StatusCode.FORBIDDEN.getStatusMessage());
+            future.complete(Boolean.FALSE);
+        }
+    }
+
+    @Override
+    public void checkConfigResource(Buffer buffer) throws ValidationException {
+        aclFactory.parseAcl(buffer);
+    }
+
+    @Override
+    public void configUpdate() {
+        storage.get(aclRoot, buffer -> {
+            if (buffer != null) {
+                grantedRoles = new HashMap<>();
+                for (Object roleObject : new JsonObject(buffer.toString("UTF-8")).getJsonArray(aclKey)) {
+                    String role = (String) roleObject;
+                    updateAcl(role);
+                }
+            } else {
+                log.warn("No ACLs in storage, using initial authorization.");
+                grantedRoles = initialGrantedRoles;
+            }
+        });
+    }
+
+
+    /**
+     * Extracts the Users Roles from the Request and further validates if the user is allowed to access
+     * the URL according to the ACL
+     * @param request The Incoming HTTP Request
+     * @return true if the user is authorized.
+     */
+    private boolean isAuthorized(HttpServerRequest request) {
+        Set<String> roles = roleExtractor.extractRoles(request);
+        if (roles != null) {
+            roles.add(anonymousRole);
+            RequestLoggerFactory.getLogger(RoleAuthorizer.class, request).debug("Roles: " + roles);
+            return isAuthorized(roles, request);
+        } else {
+            return true;
+        }
+    }
+
+
+    /**
+     * Validates that the Request URL is authorised by any of the given Roles. The roles are mapped first
+     * against the defined RoleMapper if any.
+     *
+     * @param roles The roles used to validate the ACL. Note that the given Roles
+     *              are mapped according to the RoleMapper (if any) before applied against the ACL.
+     * @param request The incoming HTTP Request
+     * @return true if the user is authorized
+     */
+    private boolean isAuthorized(Set<String> roles, HttpServerRequest request) {
+        Set<String> mappedRoles = roleMapper.mapRoles(roles);
+        for (Entry<PatternHolder, Map<String, Set<String>>> entry : grantedRoles.entrySet()) {
```

Comment:
Note that we have a unclear situation during startup:
Loading of the ACLs is async, so `grantedRoles` is filled step by step. If you have requests *while* this filling-up is in progress you can (accidentially) answer such requests with `403 FORBIDDEN` but we probably should answer with `503 Service Unavailable`

Reply:
As well here this is as it was since ages. If it is worth, please open a new ticket for this. I'll be happy to do this later in a new adventure.